### PR TITLE
[OBCQ][Recipe UX change] Prunen-prunem-to-mask-structure

### PR DESCRIPTION
**Pull request by rahul-tuli**
Before this PR OBCQ recipes needed prunen/prunem to be specified individually. This PR makes a UX change such that now, a `mask_structure` of the form `N:M` can be defined; the where both N, M are integers (similar to prunen and prunem)

Test command:

```bash
python3 src/sparseml/transformers/sparsification/obcq/obcq.py faceb
ook/opt-350m c4 --recipe src/sparseml/transformers/sparsification/obcq/example.yaml --eval wikitext2 
```

Output:
```bash
.
.
.
2023-10-31 07:03:54 sparseml.modifiers.quantization.pytorch INFO     Running quantization calibration using calibration_dataloader
2023-10-31 07:04:02 sparseml.modifiers.obcq.utils.helpers INFO     Evaluating perplexity...
2023-10-31 07:04:03 sparseml.modifiers.obcq.utils.helpers INFO     tensor(34.7602, device='cuda:0')
2023-10-31 07:04:05 sparseml.modifiers.obcq.utils.helpers INFO     tensor(36.3635, device='cuda:0')
2023-10-31 07:04:06 sparseml.modifiers.obcq.utils.helpers INFO     tensor(37.2799, device='cuda:0')
2023-10-31 07:04:08 sparseml.modifiers.obcq.utils.helpers INFO     tensor(36.4086, device='cuda:0')
2023-10-31 07:04:09 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.7602, device='cuda:0')
2023-10-31 07:04:11 sparseml.modifiers.obcq.utils.helpers INFO     tensor(35.3937, device='cuda:0')
2023-10-31 07:04:12 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.7370, device='cuda:0')
2023-10-31 07:04:14 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.7335, device='cuda:0')
2023-10-31 07:04:16 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.4511, device='cuda:0')
2023-10-31 07:04:17 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.1673, device='cuda:0')
2023-10-31 07:04:19 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.1939, device='cuda:0')
2023-10-31 07:04:20 sparseml.modifiers.obcq.utils.helpers INFO     tensor(34.0454, device='cuda:0')
2023-10-31 07:04:22 sparseml.modifiers.obcq.utils.helpers INFO     tensor(34.4664, device='cuda:0')
2023-10-31 07:04:23 sparseml.modifiers.obcq.utils.helpers INFO     tensor(34.3553, device='cuda:0')
2023-10-31 07:04:25 sparseml.modifiers.obcq.utils.helpers INFO     tensor(34.0684, device='cuda:0')
2023-10-31 07:04:26 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.9638, device='cuda:0')
2023-10-31 07:04:28 sparseml.modifiers.obcq.utils.helpers INFO     tensor(33.9543, device='cuda:0')
2023-10-31 07:04:29 sparseml.modifiers.obcq.utils.helpers INFO     tensor(34.1352, device='cuda:0')
2023-10-31 07:04:29 sparseml.modifiers.obcq.utils.helpers INFO     Perplexity: 34.135204
```

### CODE REVIEW

File: src/sparseml/modifiers/obcq/base.py [hunk 1]
```diff
@@ -37,15 +37,17 @@ class SparseGPTModifier(Modifier):
 
     :param sparsity: Sparsity to compress model to
     :param block_size: Used to determine number of columns to compress in one pass
-    :param quantize: Whether or not to quantize weights during SparseGPT. Set to True
-    to quantize using an existing quantization modifier, or pass in the configuration
-    for a quantization modifier if one does not already exist in the recipe
+    :param quantize: Whether or not to quantize weights during SparseGPT. Set to
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/obcq/base.py:40]: A cosmetic change for consistency

File: src/sparseml/modifiers/obcq/base.py [hunk 2]
```diff
@@ -135,7 +138,7 @@ def _build_quant_modifier_from_dict(self, quant_config, framework):
             **modifier_args,
         )
 
-    def _validate_layerwise_sparisity(self):
+    def _validate_layerwise_sparsity(self):
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/obcq/base.py:141]: Another cosmetic change to fix a typo

File: src/sparseml/modifiers/obcq/pytorch.py [hunk 3]
```diff
@@ -46,15 +46,15 @@ class SparseGPTModifierPyTorch(SparseGPTModifier):
 
     model: Any = None
     device_: str = "cuda:0"
-    finalization_kwargs_: Dict = None
+    finalization_kwargs_: Optional[Dict] = None
```
Comment (rahul-tuli) [inline src/sparseml/modifiers/obcq/pytorch.py:49]: Fixed Typing

**Comment by mgoin** [comment]
@rahul-tuli could you put an example of this in a recipe? especially so if this will be a breaking change to existing recipes

File: src/sparseml/modifiers/obcq/base.py [hunk 4]
```diff
@@ -58,8 +60,9 @@ class SparseGPTModifier(Modifier):
     quantize: Union[bool, Dict]
     dampening_frac: Optional[float] = 0.01
     sequential_update: Optional[bool] = True
-    prunen: Optional[int] = 0
-    prunem: Optional[int] = 0
+    mask_structure: str = "0:0"
+    prunen_: Optional[int] = None
+    prunem_: Optional[int] = None
```
Comment (dsikka) [inline src/sparseml/modifiers/obcq/base.py:65]: underscore at the end?

File: src/sparseml/transformers/sparsification/obcq/example_llama.yaml [hunk 5]
```diff
@@ -50,8 +50,7 @@ test_stage:
       sequential_update: False
       quantize: True
       percdamp: 0.01
-      prunen: 0
-      prunem: 0
+      mask_structure: "0:0"
```
Comment (dsikka) [inline src/sparseml/transformers/sparsification/obcq/example_llama.yaml:53]: Why were these changed?

File: src/sparseml/transformers/sparsification/obcq/example.yaml [hunk 6]
```diff
@@ -16,8 +16,7 @@ test_stage:
               input_activations: null
               output_activations: null
       percdamp: 0.01
-      prunen: 0
-      prunem: 0
+      mask_structure: "0:0"
```
Comment (dsikka) [inline src/sparseml/transformers/sparsification/obcq/example.yaml:19]: same as below

**Review by dsikka — APPROVED** [review]
LGTM. Just two questions since unfamiliar with this work

File: src/sparseml/modifiers/obcq/base.py (hunk 4 above)
Comment (mgoin) [inline src/sparseml/modifiers/obcq/base.py:65]: why keep both of these interfaces around if `mask_structure` must be defined? it seems currently it will clobber whatever you put in prunen or prunem? i'm questioning the justification for this change because of this complexity

Comment (rahul-tuli) [reply src/sparseml/modifiers/obcq/base.py:65]: This is the expected way to define private variables in the new framework; It's actually a very well known practice to put an underscore at the end; also Pydantic does not like underscores in the front

File: src/sparseml/transformers/sparsification/obcq/example_llama.yaml (hunk 5 above)
Comment (rahul-tuli) [reply src/sparseml/transformers/sparsification/obcq/example_llama.yaml:53]: This is the intended change; to move away from specifying prunen and prunem separately; as a future PR this mask_structure attribute can also be used to do more nifty things, like 4-block/channel/filter level masks

File: src/sparseml/transformers/sparsification/obcq/example.yaml (hunk 6 above)
Comment (rahul-tuli) [reply src/sparseml/transformers/sparsification/obcq/example.yaml:19]: Same answer as above

File: src/sparseml/modifiers/obcq/base.py (hunk 4 above)
Comment (rahul-tuli) [reply src/sparseml/modifiers/obcq/base.py:65]: Both interfaces are not around anymore, infact specifying `prunen` or `prunem` will fail; they have been converted to private variables `prunen_` and `prunem_` ; Also removed from the docstring; are you saying it can lead to confusion? if yes what are your thoughts, should we rename them somehow? or anything else you would like to see?

**Review by dsikka — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
